Add options to kubefed telling it to generate HTTP Basic and/or token credentials for the Federated API server

perotinus opened the pull request:
fixes #41265.

**Release notes**:
```release-note
Adds two options to kubefed, `-apiserver-enable-basic-auth` and `-apiserver-enable-token-auth`, which generate an HTTP Basic username/password and a token respectively for the Federated API server.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41682)
<!-- Reviewable:end -->

perotinus commented:
OK, this should be ready to review now.

madhusudancs commented:
@perotinus a few comments.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 244 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXQkRcyB5B26DcWhnx:-KdXQkRcyB5B26DcWhny:b4x5dmx) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L244)):*
> ```Go
> 	cmKubeconfigName := fmt.Sprintf("%s-kubeconfig", cmName)
> 	httpBasicAuthPassword := ""
> 	authToken := ""
> ```

Can you pull these variables into a struct? Call it `credentials` or something?

---

*[federation/pkg/kubefed/init/init.go, line 269 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXQuoyeeEEiFrC7Dvz:-KdXQuoyeeEEiFrC7Dw-:b69ry56) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L269)):*
> ```Go
> 	}
> 
> 	// 3c. Generate TLS certificates and credentials
> ```

I would put all these 3 into a function that returns the credentials struct.

---

*[federation/pkg/kubefed/init/init.go, line 590 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXS08OYNcbG6Oj7NBR:-KdXS08OYNcbG6Oj7NBS:b-bza57) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L590)):*
> ```Go
> }
> 
> func createAPIServer(clientset *client.Clientset, namespace, name, image, credentialsName string, hasHTTPBasicAuthFile, hasTokenAuthFile bool, advertiseAddress, storageBackend string, argOverrides map[string]string, pvc *api.PersistentVolumeClaim, dryRun bool) (*extensions.Deployment, error) {
> ```

Please move these bool types a little bit to the right, between `storageBackend` and `argOverrides`. You could move credentialsName also if you want them to adjacent to each other.

---

*[federation/pkg/kubefed/init/init.go, line 945 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXSiL0A3Qt5Q24lYbN:-KdXSiL1JKdPJdt1CSCn:b-x8gcly) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L945)):*
> ```Go
> 	cluster.CertificateAuthorityData = certutil.EncodeCertPEM(entKeyPairs.ca.Cert)
> 
> 	var httpBasicAuthInfo *clientcmdapi.AuthInfo
> ```

Move this below, closer to its initialization.

---

*[federation/pkg/kubefed/init/init.go, line 984 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXT1ybcR_IAGNDhe_s:-KdXT1ybcR_IAGNDhe_t:bepsfij) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L984)):*
> ```Go
> 
> // authFileContents returns a CSV string containing the contents of an
> // authenticaion file in the format required by the federation-apiserver.
> ```

s/authenticaion/authentication/

---

*[federation/pkg/kubefed/init/init_test.go, line 1087 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXgIdlF3A3iRZtQMGy:-KdXgIdlF3A3iRZtQMGz:b-rblcdm) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1087)):*
> ```Go
> 
> 				// Verify that the secret contains the HTTP Basic and token credentials.
> 				if got.Name == credSecretName {
> ```

There is a switch below for exactly for this case. I see you have this if block here because we nullify `got.Data`. You can move that line along with its comment to follow the `switch` statement.

---

*[federation/pkg/kubefed/init/init_test.go, line 1088 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXh2JpSbM0Pi5fcJXa:-KdXh2JpSbM0Pi5fcJXb:bwpgbys) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1088)):*
> ```Go
> 				// Verify that the secret contains the HTTP Basic and token credentials.
> 				if got.Name == credSecretName {
> 					if apiserverEnableHTTPBasicAuth && got.Data["basicauth.csv"] == nil {
> ```

Also, I think this will be more readable if you collapse this into a nested if-else block

```golang
if apiserverEnableHTTPBasicAuth {
    if got.Data["basicauth.csv"] == nil {
        ...
    }
    if got.Data["token.csv"] == nil {
        ...
    }
} else {
    if got.Data["basicauth.csv"] != nil {
        ...
    }
    if got.Data["token.csv"] != nil {
        ...
    }
}
```

---

*[federation/pkg/kubefed/init/init_test.go, line 1195 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXlpHhMp0sSp43UbNr:-KdXlpHhMp0sSp43UbNs:b-kxtsxq) ([raw file](https://github.com/kubernetes/kubernetes/blob/b201ac2f8f8328d9c828f36358b106d0ea21a14a/federation/pkg/kubefed/init/init_test.go#L1195)):*
> ```Go
> 	}
> 	if authInfo.Username != AdminCN {
> 		t.Errorf("Want username: %q, got: %q", AdminCN, authInfo.Username)
> ```

Why remove this?

---

*[federation/pkg/kubefed/init/init_test.go, line 1242 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXlxw7v0xZOlumeJ0m:-KdXlxw7v0xZOlumeJ0n:bl6c02y) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1242)):*
> ```Go
> 	}
> 	if !apiserverEnableTokenAuth && len(authInfo.Token) != 0 {
> 		t.Errorf("Expected token to be empty")
> ```

Also, print what you got. Convention is to write this string as `t.Errorf("Expected token to be empty: got: %s", authInfo.Token)` 

---

*[federation/pkg/kubefed/init/init_test.go, line 1245 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXmIkHAv4lpq9D_S60:-KdXmIkHAv4lpq9D_S61:b-4r2n4l) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1245)):*
> ```Go
> 	}
> 	if apiserverEnableTokenAuth && len(authInfo.Token) == 0 {
> 		t.Errorf("Expected token to be non-empty")
> ```

Print what you want. Convention is to write this string as `t.Errorf("Expected token to be non-empty: want: %s", authInfo.Token)` 

---

*[federation/pkg/kubefed/init/init_test.go, line 1258 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXmYtyod-AHysAVd-9:-KdXmYtyod-AHysAVd-A:benpsw9) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1258)):*
> ```Go
> 		}
> 		if len(httpBasicAuthInfo.Username) == 0 {
> 			t.Errorf("Expected basic auth AuthInfo entry to contain username")
> ```

Same comment as above.

---

*[federation/pkg/kubefed/init/init_test.go, line 1261 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXm_XbuWoA25MBYhBh:-KdXm_XbuWoA25MBYhBi:b2nreh6) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1261)):*
> ```Go
> 		}
> 		if len(httpBasicAuthInfo.Password) == 0 {
> 			t.Errorf("Expected basic auth AuthInfo entry to contain password")
> ```

Same as above.

---

*[test/e2e_federation/federation-authn.go, line 45 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXpsFug2SAG0GFCt6Y:-KdXpsFvioR4FpH1dM1D:bqs6jzd) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/test/e2e_federation/federation-authn.go#L45)):*
> ```Go
> 
> // validity specifies whether a credential should be valid or invalid.
> type validity int
> ```

Can't this be a simple bool?

---

*[test/e2e_federation/federation-authn.go, line 135 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXp8JrtCRT32TPw-qm:-KdXp8JrtCRT32TPw-qn:b-ngcv8b) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/test/e2e_federation/federation-authn.go#L135)):*
> ```Go
> // federationClientSet returns a Federation Clientset configured to authenticate
> // with the given credential type and validity.
> func federationClientSet(authType credentialType, validity validity) (*federation_clientset.Clientset, error) {
> ```

There is a lot of logic in this code. Also see https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html.

Could we please split this into 3 separate simple functions: `federationClientsetWithCert`, `federationClientsetWithToken` and `federationClientSetWithPW`? Or something like that.

Also, I am not sure about using `validity` bool to create invalid credentials, but let it be there for now. We can refactor things later if it feels like too much logic or when we move these tests to integration tests.

---

*[test/e2e_federation/federation-authn.go, line 175 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXqYEuq6t8SgsAlvgb:-KdXqYEuq6t8SgsAlvgc:b6qibum) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/test/e2e_federation/federation-authn.go#L175)):*
> ```Go
> 			// This is a hacky approach to getting the basic auth credentials, but since
> 			// the token and the username/password cannot live in the same AuthInfo object,
> 			// we must dig it out by hand.
> ```

It's not just that, we also don't want to keep basic auth with anything else, even if there weren't tokens in the same object.

---

*[test/e2e_federation/federation-authn.go, line 181 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXqqk_P11eohBXNmwy:-KdXqqk_P11eohBXNmwz:b4ksjin) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/test/e2e_federation/federation-authn.go#L181)):*
> ```Go
> 			}
> 			for name, authInfo := range c.AuthInfos {
> 				if strings.HasPrefix(name, framework.TestContext.FederatedKubeContext) && strings.HasSuffix(name, "basic-auth") {
> ```

You should be able to just lookup for `c.AuthInfos[fmt.Sprintf("%s-basic-auth", framework.TestContext.FederatedKubeContext, )]` right? That looks more readable to me.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-:-KdXr9FglNKoHmOOh6Dv:b-2ud2g3)*
<!-- Sent from Reviewable.io -->

perotinus commented:
> from perotinus/kubernetes:unpwandtokens
>             
>             to kubernetes/kubernetes:master
Thanks a lot for the great comments!

---

Review status: 0 of 3 files reviewed at latest revision, 16 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 244 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXQkRcyB5B26DcWhnx:-KdYFQ2MK1Pb7I-UjTuB:b-va1dku) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L244)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Can you pull these variables into a struct? Call it `credentials` or something?
</blockquote></details>

Done, I believe. I pulled the entity key pairs in as well for good measure.

---

*[federation/pkg/kubefed/init/init.go, line 269 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXQuoyeeEEiFrC7Dvz:-KdYFXB3c_SpTC4YR3wu:b7rpr8b) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L269)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I would put all these 3 into a function that returns the credentials struct.
</blockquote></details>

Good call. I've added a new function that calls the three previously-existing ones. WDYT?

---

*[federation/pkg/kubefed/init/init.go, line 590 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXS08OYNcbG6Oj7NBR:-KdYFanN75czmYfcb7fD:b-am9vsa) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L590)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Please move these bool types a little bit to the right, between `storageBackend` and `argOverrides`. You could move credentialsName also if you want them to adjacent to each other.
</blockquote></details>

Thanks! I wasn't quite sure where to put them, but just before the overrides makes a lot of sense.

---

*[federation/pkg/kubefed/init/init.go, line 945 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXSiL0A3Qt5Q24lYbN:-KdYFhnm3OEjyPhD0ljq:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L945)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Move this below, closer to its initialization.
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init.go, line 984 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXT1ybcR_IAGNDhe_s:-KdYFi7HYfKSSuE8zUqV:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L984)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

s/authenticaion/authentication/
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init_test.go, line 1087 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXgIdlF3A3iRZtQMGy:-KdYFjalpUZcKlwW7snr:bukchya) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1087)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

There is a switch below for exactly for this case. I see you have this if block here because we nullify `got.Data`. You can move that line along with its comment to follow the `switch` statement.
</blockquote></details>

Ah, of course! Thanks.

---

*[federation/pkg/kubefed/init/init_test.go, line 1088 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXh2JpSbM0Pi5fcJXa:-KdYFlCHFa0pcVtmkaS6:bqnyo2d) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1088)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Also, I think this will be more readable if you collapse this into a nested if-else block

```golang
if apiserverEnableHTTPBasicAuth {
    if got.Data["basicauth.csv"] == nil {
        ...
    }
    if got.Data["token.csv"] == nil {
        ...
    }
} else {
    if got.Data["basicauth.csv"] != nil {
        ...
    }
    if got.Data["token.csv"] != nil {
        ...
    }
}
```
</blockquote></details>

Done, in a manner of speaking. There is only one conditional inside each `apiserverEnableX` branch, since each of those flags only affects one of the files.

---

*[federation/pkg/kubefed/init/init_test.go, line 1195 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXlpHhMp0sSp43UbNr:-KdYGJM7hKBwEMvwim3B:bt54g60) ([raw file](https://github.com/kubernetes/kubernetes/blob/b201ac2f8f8328d9c828f36358b106d0ea21a14a/federation/pkg/kubefed/init/init_test.go#L1195)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Why remove this?
</blockquote></details>

This is actually due to a real change in kubefed; the username and token cannot both be populated in the 

---

*[federation/pkg/kubefed/init/init_test.go, line 1242 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXlxw7v0xZOlumeJ0m:-KdYFyWjccPjBtozSfDM:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1242)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Also, print what you got. Convention is to write this string as `t.Errorf("Expected token to be empty: got: %s", authInfo.Token)` 
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init_test.go, line 1245 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXmIkHAv4lpq9D_S60:-KdYFzZArBwvN1JxCR69:b9w7chl) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1245)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Print what you want. Convention is to write this string as `t.Errorf("Expected token to be non-empty: want: %s", authInfo.Token)` 
</blockquote></details>

Hmm. That makes sense, but how do I get this? The want value is a generated value that I can't necessarily get access to, right? If it were there, the test would pass.

---

*[test/e2e_federation/federation-authn.go, line 45 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXpsFug2SAG0GFCt6Y:-KdYGVct7qU1LfO4S_dk:b-8wujcq) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/test/e2e_federation/federation-authn.go#L45)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Can't this be a simple bool?
</blockquote></details>

Yes. Yes it can.

---

*[test/e2e_federation/federation-authn.go, line 135 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXp8JrtCRT32TPw-qm:-KdYGXU4LEVPFtD6tvRQ:b-rmbmuz) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/test/e2e_federation/federation-authn.go#L135)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

There is a lot of logic in this code. Also see https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html.

Could we please split this into 3 separate simple functions: `federationClientsetWithCert`, `federationClientsetWithToken` and `federationClientSetWithPW`? Or something like that.

Also, I am not sure about using `validity` bool to create invalid credentials, but let it be there for now. We can refactor things later if it feels like too much logic or when we move these tests to integration tests.
</blockquote></details>

Done. More code, but definitely clearer.

---

*[test/e2e_federation/federation-authn.go, line 175 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXqYEuq6t8SgsAlvgb:-KdYGZiqNkO2-PnB6RZN:b-9nvo58) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/test/e2e_federation/federation-authn.go#L175)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

It's not just that, we also don't want to keep basic auth with anything else, even if there weren't tokens in the same object.
</blockquote></details>

Ah, yes. I've added that to the comment.

---

*[test/e2e_federation/federation-authn.go, line 181 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXqqk_P11eohBXNmwy:-KdYGe3C1gMIwmxPSsGR:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/test/e2e_federation/federation-authn.go#L181)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

You should be able to just lookup for `c.AuthInfos[fmt.Sprintf("%s-basic-auth", framework.TestContext.FederatedKubeContext, )]` right? That looks more readable to me.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-:-KdYQr35LTOBQxohcS5o:bc1oeql)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: 0 of 3 files reviewed at latest revision, 16 unresolved discussions.

---

*[federation/pkg/kubefed/init/init_test.go, line 1258 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXmYtyod-AHysAVd-9:-KdYQz1bnUyX5sI355e-:bm0ggup) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1258)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Same comment as above.
</blockquote></details>

Thanks! See my question above.

---

*[federation/pkg/kubefed/init/init_test.go, line 1261 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXm_XbuWoA25MBYhBh:-KdYR2LqQ-hUGQwbFJxA:b-4kl4tk) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1261)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

Also see my question above.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@perotinus a few minor nits.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 244 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXQkRcyB5B26DcWhnx:-KdeLgp1LhsDkuU7Wzmr:biygx3j) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L244)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Done, I believe. I pulled the entity key pairs in as well for good measure.
</blockquote></details>

Yeah, I actually meant to pull all of them, but did not make it clear. Sorry.

---

*[federation/pkg/kubefed/init/init.go, line 269 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXQuoyeeEEiFrC7Dvz:-KdeMIQLxgHXlE3YZfhG:b32w49) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L269)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Good call. I've added a new function that calls the three previously-existing ones. WDYT?
</blockquote></details>

SG

---

*[federation/pkg/kubefed/init/init.go, line 493 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdeMmE0yB22gQ5MlYMK:-KdeMmE0yB22gQ5MlYML:b6izzrc) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/federation/pkg/kubefed/init/init.go#L493)):*
> ```Go
> 	credentials.certEntKeyPairs = entKeyPairs
> 
> 	_, err = createAPIServerCredentialsSecret(clientset, svcNamespace, serverCredName, &credentials, dryRun)
> ```

I think this call shouldn't be inside this function. You are calling the function `generateCredentials()`, but creating credentials secret here. It makes sense to have this call in the top-level function after the call to `generateCredentials()`

---

*[federation/pkg/kubefed/init/init.go, line 554 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdeOIpFdP3LH9oKQZG3:-KdeOIpFdP3LH9oKQZG4:b-eydk7h) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/federation/pkg/kubefed/init/init.go#L554)):*
> ```Go
> }
> 
> func createControllerManagerKubeconfigSecret(clientset *client.Clientset, namespace, name, svcName, kubeconfigName string, credentials *credentials, dryRun bool) (*api.Secret, error) {
> ```

Is it necessary to pass the entire `credentials` struct here? Isn't it sufficient to just pass `credentials.certEntKeyPairs`?

---

*[federation/pkg/kubefed/init/init_test.go, line 1088 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXh2JpSbM0Pi5fcJXa:-Kdj7hcJn6PV6nj4bGb6:bw0hlan) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1088)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Done, in a manner of speaking. There is only one conditional inside each `apiserverEnableX` branch, since each of those flags only affects one of the files.
</blockquote></details>

Makes sense. Looks good now.

---

*[federation/pkg/kubefed/init/init_test.go, line 1195 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXlpHhMp0sSp43UbNr:-Kdj9_VT863LpQXd6bZu:b-m6zal7) ([raw file](https://github.com/kubernetes/kubernetes/blob/b201ac2f8f8328d9c828f36358b106d0ea21a14a/federation/pkg/kubefed/init/init_test.go#L1195)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

This is actually due to a real change in kubefed; the username and token cannot both be populated in the 
</blockquote></details>

Fair enough. But what I meant was, this should now be a test for the basic auth entry instead. See my other comment above about username being a static string.

---

*[federation/pkg/kubefed/init/init_test.go, line 1245 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXmIkHAv4lpq9D_S60:-KdeQs3ImVGalJL2C4fV:b-70m4kv) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1245)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Hmm. That makes sense, but how do I get this? The want value is a generated value that I can't necessarily get access to, right? If it were there, the test would pass.
</blockquote></details>

Yeah, ignore my original comment. For a moment here, I missed the fact that this is a generated string.

---

*[federation/pkg/kubefed/init/init_test.go, line 1258 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXmYtyod-AHysAVd-9:-Kdj87x6isKxbMe1H9e5:ba9w2xx) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1258)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Thanks! See my question above.
</blockquote></details>

Username is a fixed/static string in your code. So you know what you want here? If you were generating this, then there wouldn't be a way. But that's not the case here, right?

---

*[federation/pkg/kubefed/init/init_test.go, line 1261 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXm_XbuWoA25MBYhBh:-Kdj8MMR8fOelDlhGZp-:b-ljzdho) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1261)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Also see my question above.
</blockquote></details>

Password is generated, you can't do much about it. Ignore my comment.

---

*[federation/pkg/kubefed/init/init_test.go, line 1091 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdePVIChzDS-4Ut9Dyx:-KdePVIChzDS-4Ut9Dyy:bnbi52a) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/federation/pkg/kubefed/init/init_test.go#L1091)):*
> ```Go
> 					if apiserverEnableHTTPBasicAuth {
> 						if got.Data["basicauth.csv"] == nil {
> 							return nil, fmt.Errorf("Expected secret data key 'basicauth.csv', but got nil")
> ```

Go's convention is to begin error strings in `fmt.Errorf()` with lower case letters, as in, in non-capitalized forms. Because these strings are usually tacked on to other strings which are finally logged. For example this function's caller does something like this:

```go
_, err := fakeInitHostFactory()
if err != nil {
    t.Errorf("Failed to intialize host factory: %v", err)
}
```

Note the capitalization here while logging and the distinction between logging and returning errors with messages.

---

*[test/e2e_federation/federation-authn.go, line 165 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdjCCvosnLawFTZb3Bj:-KdjCCvosnLawFTZb3Bk:b-yrkhr4) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/test/e2e_federation/federation-authn.go#L165)):*
> ```Go
> 	config.BearerToken = ""
> 
> 	if validity == valid {
> ```

You mean `validity == invalid`?

This is one of the confusions about using an equivalent of enum for truly boolean values. This is much more clear and leads to little confusion if you just get rid of the validity type alias and just define your function as `federationClientSetWithBasicAuth(valid bool) ... `. Then in your caller, you can say `federationClientSetWithBasicAuth(false /* invalid */)` or some such. But here in this block all you have to do is:

```go
if valid {
   // Hacky approach to get basic auth credentials
} else {
   // empty/invalid strings
}
```

The problem here is also kind of compounded by the fact that the parameter name and the type name in the function definition are both called `validity`.

---

*[test/e2e_federation/federation-authn.go, line 205 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdjGbPOtYDnEGjccdFF:-KdjGbPOtYDnEGjccdFG:b-wsbmz) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/test/e2e_federation/federation-authn.go#L205)):*
> ```Go
> 	config.Password = ""
> 
> 	if validity == invalid {
> ```

Also here. It is simple to just use a bool instead of this type alias.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-:-KdjGkPlA4UfzXVJ4hUq:bckf3qk)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Thanks again! PTAL.

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 244 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXQkRcyB5B26DcWhnx:-KdjNFtNwkNb04frS9T9:b-vqkjus) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L244)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Yeah, I actually meant to pull all of them, but did not make it clear. Sorry.
</blockquote></details>

No worries!

---

*[federation/pkg/kubefed/init/init.go, line 269 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXQuoyeeEEiFrC7Dvz:-KdjNILgG_InJuJqcP_S:b-ucfknh) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init.go#L269)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

SG
</blockquote></details>

Thanks!

---

*[federation/pkg/kubefed/init/init.go, line 493 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdeMmE0yB22gQ5MlYMK:-KdjKBBbitWrUUaxf0pH:b-3b5juu) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/federation/pkg/kubefed/init/init.go#L493)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I think this call shouldn't be inside this function. You are calling the function `generateCredentials()`, but creating credentials secret here. It makes sense to have this call in the top-level function after the call to `generateCredentials()`
</blockquote></details>

Ah, I see your point. That makes sense. Done!

---

*[federation/pkg/kubefed/init/init.go, line 554 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdeOIpFdP3LH9oKQZG3:-KdjKU0RI8c2if0LMcF9:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/federation/pkg/kubefed/init/init.go#L554)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Is it necessary to pass the entire `credentials` struct here? Isn't it sufficient to just pass `credentials.certEntKeyPairs`?
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init_test.go, line 1088 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXh2JpSbM0Pi5fcJXa:-KdjKVrrINzP1wYmAogc:b-ucfknh) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1088)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Makes sense. Looks good now.
</blockquote></details>

Thanks!

---

*[federation/pkg/kubefed/init/init_test.go, line 1195 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXlpHhMp0sSp43UbNr:-KdjLZqgEFp-7g2rQg-K:bu0tvpv) ([raw file](https://github.com/kubernetes/kubernetes/blob/b201ac2f8f8328d9c828f36358b106d0ea21a14a/federation/pkg/kubefed/init/init_test.go#L1195)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Fair enough. But what I meant was, this should now be a test for the basic auth entry instead. See my other comment above about username being a static string.
</blockquote></details>

Hmm. I'm not quite sure I follow: this change was due to the fact that the original, already-existing AuthInfo that contained the cert now contains a token, and so cannot contain a username. There's another test below that tests the HTTP Basic AuthInfo object, so I do test that below. Is that what you meant by a test for the basic auth entry?

---

*[federation/pkg/kubefed/init/init_test.go, line 1245 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXmIkHAv4lpq9D_S60:-KdjNJMztLn3DF42yBoN:b-vqkjus) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1245)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Yeah, ignore my original comment. For a moment here, I missed the fact that this is a generated string.
</blockquote></details>

No worries!

---

*[federation/pkg/kubefed/init/init_test.go, line 1258 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXmYtyod-AHysAVd-9:-KdjMMXs6FGvTU8S_WZb:b-ailise) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1258)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Username is a fixed/static string in your code. So you know what you want here? If you were generating this, then there wouldn't be a way. But that's not the case here, right?
</blockquote></details>

Ah, yes, that's absolutely true. Done.

---

*[federation/pkg/kubefed/init/init_test.go, line 1261 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXm_XbuWoA25MBYhBh:-KdjNO67epbP76gMMfUy:bb6sh2i) ([raw file](https://github.com/kubernetes/kubernetes/blob/bf00e54e562d775c9bebf0ed9be88787a82b3266/federation/pkg/kubefed/init/init_test.go#L1261)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Password is generated, you can't do much about it. Ignore my comment.
</blockquote></details>

Ack.

---

*[federation/pkg/kubefed/init/init_test.go, line 1091 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdePVIChzDS-4Ut9Dyx:-KdjKXsiv-cI23mVvMiE:bnt78yt) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/federation/pkg/kubefed/init/init_test.go#L1091)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Go's convention is to begin error strings in `fmt.Errorf()` with lower case letters, as in, in non-capitalized forms. Because these strings are usually tacked on to other strings which are finally logged. For example this function's caller does something like this:

```go
_, err := fakeInitHostFactory()
if err != nil {
    t.Errorf("Failed to intialize host factory: %v", err)
}
```

Note the capitalization here while logging and the distinction between logging and returning errors with messages.


</blockquote></details>

Ah, yes. This convention hasn't worked it's way into my head yet. Fixed (here and elsewhere).

---

*[test/e2e_federation/federation-authn.go, line 165 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdjCCvosnLawFTZb3Bj:-KdjN34Y6lsKOHwoRbTo:b-jf44yb) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/test/e2e_federation/federation-authn.go#L165)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

You mean `validity == invalid`?

This is one of the confusions about using an equivalent of enum for truly boolean values. This is much more clear and leads to little confusion if you just get rid of the validity type alias and just define your function as `federationClientSetWithBasicAuth(valid bool) ... `. Then in your caller, you can say `federationClientSetWithBasicAuth(false /* invalid */)` or some such. But here in this block all you have to do is:

```go
if valid {
   // Hacky approach to get basic auth credentials
} else {
   // empty/invalid strings
}
```

The problem here is also kind of compounded by the fact that the parameter name and the type name in the function definition are both called `validity`.
</blockquote></details>

Yeah, you're right. Now that I look at it again, the bool is a lot clearer. Fixed.

---

*[test/e2e_federation/federation-authn.go, line 205 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdjGbPOtYDnEGjccdFF:-KdjN8NNG16d6n23Dc-p:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/test/e2e_federation/federation-authn.go#L205)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Also here. It is simple to just use a bool instead of this type alias.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-:-KdjND7TjHZbVc63yh9t:bojy6b6)*
<!-- Sent from Reviewable.io -->

perotinus commented:
@madhusudancs Ping?

madhusudancs commented:
Needs one more fix.

---

Reviewed 2 of 3 files at r3.
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/kubefed/init/init_test.go, line 1195 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdXlpHhMp0sSp43UbNr:-KdoPZFEXzlwYwvbZ4nI:bbiy45m) ([raw file](https://github.com/kubernetes/kubernetes/blob/b201ac2f8f8328d9c828f36358b106d0ea21a14a/federation/pkg/kubefed/init/init_test.go#L1195)):*
> Is that what you meant by a test for the basic auth entry?

Yes.

---

*[test/e2e_federation/federation-authn.go, line 165 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdjCCvosnLawFTZb3Bj:-KdoQ4clQN0q6Ee0tXg9:b-z5k2pv) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/test/e2e_federation/federation-authn.go#L165)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Yeah, you're right. Now that I look at it again, the bool is a lot clearer. Fixed.
</blockquote></details>

I am still not sure if this is right. Shouldn't you swap your condition or the corresponding blocks? See the example in the previous comment.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-:-KdoQSb9BUULh_R_0dIq:b-2leihq)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Reviewed 1 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Thanks! PTAL.

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/e2e_federation/federation-authn.go, line 165 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdjCCvosnLawFTZb3Bj:-KdtN5h3XzZFTkiMXFVN:b5fxdf4) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/test/e2e_federation/federation-authn.go#L165)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I am still not sure if this is right. Shouldn't you swap your condition or the corresponding blocks? See the example in the previous comment.
</blockquote></details>

Hrm. You're right, this should be !valid. Fixed.

The e2e tests are currently failing to bring up a federation API server for me, so I haven't been able to run them again to verify.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-:-KdtNIVKl3oMx0pIWS5H:b4e8i7m)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
/lgtm
/approve

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[test/e2e_federation/federation-authn.go, line 165 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-KdjCCvosnLawFTZb3Bj:-KdtjWP01wloAZVjSQmi:bd97r8x) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3687a5339a8781bc46d7f5d47cc702962588cab/test/e2e_federation/federation-authn.go#L165)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Hrm. You're right, this should be !valid. Fixed.

The e2e tests are currently failing to bring up a federation API server for me, so I haven't been able to run them again to verify.
</blockquote></details>

If you rebase to master-HEAD now, you should be able to run the federation e2e tests.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41682#-:-KdtjqVTihjj9BHx9RY4:b6yv7yi)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@perotinus it looks like BUILD files need an update. Please run `hack/update-bazel.sh`.

madhusudancs commented:
@k8s-bot gci gce e2e test this

madhusudancs commented:
@perotinus that cmd/BUILD change is strange. I don't understand. May be rebase again and try running hack/bazel-update.sh?

perotinus commented:
@madhusudancs Yeah, I noticed that too and was fixing it just as you commented. It should be better now.

perotinus commented:
@k8s-bot gci gce e2e test this

perotinus commented:
@madhusudancs Can you poke the robot here too? Thanks!

perotinus committed:
Adds support for HTTP basic and token authentication to kubefed.

perotinus commented:
@madhusudancs This may need a re-LGTM: I managed to run the e2e tests again, and modified federation-up.sh to unbreak two of the tests.

csbell commented:
@k8s-bot gci gce e2e test this

madhusudancs commented on federation/cluster/federation-up.sh:
```diff
@@ -82,7 +82,9 @@ function init() {
       --dns-zone-name="${DNS_ZONE_NAME}" \
-      --image="${kube_registry}/hyperkube-amd64:${kube_version}"
+      --image="${kube_registry}/hyperkube-amd64:${kube_version}" \
+      --apiserver-enable-basic-auth \
```
Can you please explicitly say `=true` here and in the next command?

perotinus replied:
Done!

madhusudancs reviewed (commented):
@perotinus Reviewable isn't working for me. So I am reviewing here on Github. Sorry about the mix-up though.

perotinus commented:
@madhusudancs Does this need another robot poke? The checks are all waiting for status to be reported.

madhusudancs commented:
/lgtm

madhusudancs commented:
@perotinus yeah, bot needs a poke. Adding an lgtm/approval works too. I did some branch prediction and applied the LGTM label ;-)

madhusudancs commented:
@perotinus looks like branch prediction failed ;-) Bazel failure looks like a legitimate failure. Could you please fix it?

perotinus commented:
@madhusudancs Failure should be fixed now. Sorry about that!

madhusudancs commented:
/lgtm

madhusudancs commented:
@perotinus np.

madhusudancs commented:
/lgtm

madhusudancs commented:
@k8s-bot verify test this

madhusudancs commented:
@perotinus this is a legitimate failure. Please fix.

perotinus commented:
cc @mwielgus @brendandburns Can you approve the changes to known-flags.txt? Also, are there supposed to be multiple entries for one flag? It seems like there are some duplicated entries that were added.

perotinus commented:
@madhusudancs Fixes made, but it looks like this needs additional approval now.

janetkuo commented:
/approve

madhusudancs commented:
@janetkuo thanks!

madhusudancs commented:
@k8s-bot ok to test

madhusudancs commented:
@kubernetes/release-admins This PR is not critical for v1.6 but it will be a useful feature. This PR was LGTM'ed a few days ago but was struck in a rebase hell. I forgot to add the milestone when I LGTM'ed the PR. Can we get this into 1.6?

madhusudancs commented:
@k8s-bot kubemark e2e test this

perotinus commented:
@madhusudancs Can you LGTM this again? I had to do another rebase.

madhusudancs commented:
@k8s-bot cvm gce e2e test this

madhusudancs commented:
/lgtm
